tests/test_api.py

```diff
@@ -175,7 +224,28 @@ def test_save_errors() -> None:
         zarr.save("data/example.zarr", a, mode="w")
 
 
-def test_open_with_mode_r(tmp_path: pathlib.Path) -> None:
+@pytest.mark.parametrize(
+    "store",
+    ["local", "memory", "remote", pytest.param("zip", marks=pytest.mark.xfail)],
+    indirect=True,
+)
+def test_open_store_with_mode_r(store: Store) -> None:
+    # 'r' means read only (must exist)
+    with pytest.raises(FileNotFoundError):
+        zarr.open(store=store, mode="r")
+
+    z1 = zarr.ones(store=store, shape=(3, 3))
+    assert z1.fill_value == 1
+
+    z2 = zarr.open(store=store, mode="r")
+    assert isinstance(z2, Array)
+    assert z2.fill_value == 1
+    assert (z2[:] == 1).all()
+    with pytest.raises(ValueError):
+        z2[:] = 3
```

Comment:
@jhamman this test is failing in this PR. What's the expected behavior here, given the new semantics of `read_only` for stores? Should we raise an error in `zarr.open` if the mode is `r` but the store is not read only? 

Note the test below, which is copied from `main`, passes, but in that case `zarr.open` takes a path / string and thus has the opportunity to create the store in read only mode.

Reply:
see https://github.com/zarr-developers/zarr-python/pull/2447#issuecomment-2500205184